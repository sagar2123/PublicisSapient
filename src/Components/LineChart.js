import React from "react";
import { Chart } from "react-google-charts";

export const LineChart = ({data, xTitle, yTitle}) => {
    return (
        <div className="chart-container">
            <Chart
                height={'400px'}
                chartType="LineChart"
                loader={<div>Loading Chart</div>}
                data={data}
                options={{
                    hAxis: {
                        title: xTitle,
                    },
                    vAxis: {
                        title: yTitle,
                    },
                    backgroundColor: "f6f6f0"
                }}
                rootProps={{ 'data-testid': '1' }}
                />
        </div>
    )
}