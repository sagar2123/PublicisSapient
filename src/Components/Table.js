import React from 'react';
import { getDataFromLocalStorage ,setDataToLocalStorage } from '../Utility/utils';

export const Table = ({data, headers, updateElementAtIndex, hideElement}) => {

    const createHeaders = () => {
        return headers.map((header, index) => {
            return <th key={header+index}>{header}</th>;
        });
    };

    const setCurrentDataInLocalStorage = (operation, currentData, index) => {
        let updatedData = JSON.parse(JSON.stringify(currentData));
        if(operation === "inc"){
            updatedData.points = parseInt(updatedData.points) + 1; 
            updatedData.voted = true;
        } else {
            updatedData.points = parseInt(updatedData.points) - 1; 
            updatedData.voted = false;
        }

        const checkIfAlreadyExistInLocalStorage = getDataFromLocalStorage(currentData.objectID);
        if(checkIfAlreadyExistInLocalStorage){
            setDataToLocalStorage({...checkIfAlreadyExistInLocalStorage, ...updatedData});
        } else {
            setDataToLocalStorage({...updatedData});
        }
        updateElementAtIndex(index, updatedData);
    };

    const createRows = () => {
        return data.map((row, index) => {
            return (
                <tr key={row.objectID}>
                    <td className="text-small align-center">{row.num_comments || "-"}</td>
                    <td className="text-small align-center">{row.points || "-"}</td>
                    <td className="text-small align-center">{
                        !row.voted ?
                        <div onClick={() => {setCurrentDataInLocalStorage("inc", row, index)}} className={"triangle-up-icon cursor-pointer"}></div> : 
                        <div onClick={() => {setCurrentDataInLocalStorage("dec", row, index)}} className={"triangle-down-icon cursor-pointer"}></div>
                    }</td>
                    <td className="text-small">{row.title || "-"} by {row.author || "-"} 
                    <button className={"text-small float-right hide-button btn-transparent cursor-pointer"} onClick={() => {hideElement(index)}}>hide</button>
                    </td>
                </tr>
            );
        });
    }

    return(
        <table>
            <thead>
                <tr>
                    {createHeaders()}
                </tr>
            </thead>
            <tbody>
                {createRows()}
            </tbody>
        </table>
    )
}

