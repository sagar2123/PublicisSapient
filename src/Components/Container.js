import React, { useEffect, useState } from 'react';
import { getPageData } from "../Service/apiFactory";
import { getDataFromLocalStorage } from "../Utility/utils";
import { LineChart } from "./LineChart";
import { Table } from './Table';

export const Container =  (props) => {
    const [data, setData] = useState([]);
    
    useEffect(() => {
        loadData();
        if(!props.match.params.pageId){
            props.history.push(`/page/1`);
        }
    }, [props.match.params.pageId]);
    
    console.log("client params", props.match)
    const loadData = async () => {
        const result = await getPageData(props.match.params.pageId);
        const currentData = result.data.hits.map((currentObj) => {
            const currentDataInStorage = getDataFromLocalStorage(currentObj.objectID);
            const currentDataObj = {
                points: currentObj.points,
                title: currentObj.title,
                author: currentObj.author,
                objectID: currentObj.objectID,
                url: currentObj.url,
                num_comments: currentObj.num_comments,
                voted: false
            }
            return currentDataInStorage ? {
                ...currentDataObj,
                ...currentDataInStorage
            } : {
                ...currentDataObj
            }
        });
        setData(currentData);
    };
    

    const updateElementAtParticularIndex = (index, currentIndexData) => {
        let updatedData = JSON.parse(JSON.stringify(data));
        updatedData[index] = {...updatedData[index], ...currentIndexData};
        setData(updatedData);
    }

    const changePage = (operation) => {
        if(operation === "inc"){
            props.history.push(`/page/${parseInt(props.match.params.pageId) + 1}`);
        } else {
            props.history.push(props.match.params.pageId <= 1 ? `/page/1` : `/page/${parseInt(props.match.params.pageId) - 1}`);
        }
    };

    const prepareLineChartData = (data) => {
        let finalData = data.map((dataObj) => {
            return [dataObj.objectID, dataObj.points]
        });
        return [["Id", "Votes"], ...finalData];
    };

    const hideElement = (index) => {
        let updatedData = JSON.parse(JSON.stringify(data));
        updatedData.splice(index, 1);
        setData(updatedData);
    }

    return (
        <div className="container">
            <>
                <Table headers={["Comments", "Vote Count", "Up Vote", "News Details"]} updateElementAtIndex={updateElementAtParticularIndex} hideElement={hideElement} data={data}/>
                <div className="page-controls background-primary">
                    <div className="app-primary-color">
                        <button className="btn-transparent app-primary-color cursor-pointer" onClick={() => {changePage("dec")}}>Previous</button>
                        | 
                        <button className="btn-transparent app-primary-color cursor-pointer" onClick={() => {changePage("inc")}}>Next</button>
                    </div>
                </div>
                <LineChart data={data.length > 0 ? prepareLineChartData(data) : []} xTitle={"Id"} yTitle={"Votes"}/>
            </> 
        </div>
    )
}