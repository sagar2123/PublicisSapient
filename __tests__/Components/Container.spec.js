import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Container } from "../../src/Components/Container";

const defaultProps = {
    match: {
        params: {
            pageId: 1
        }
    }
}

describe("<Container />", () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(
			<Container {...defaultProps}/>
		);
	});

	it("renders without crashing and updates the state", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
	});
});
