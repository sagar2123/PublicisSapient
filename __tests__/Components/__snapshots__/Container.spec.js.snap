// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> renders without crashing and updates the state 1`] = `
<div
  className="container"
>
  <div>
    Loading...
  </div>
</div>
`;
